File: contrib/python/bareagent.py
"""bareagent Python wrapper — subprocess + JSONL over stdin/stdout."""

import json
import os
import subprocess


class BareAgent:
    """Thin wrapper that spawns bareagent as a subprocess."""

    def __init__(self, provider="openai", model=None, url=None):
        cmd = ["npx", "bareagent", "--jsonl", "--provider", provider]
        if model:
            cmd += ["--model", model]
        if url:
            cmd += ["--url", url]
        self.proc = subprocess.Popen(
            cmd,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            env={**os.environ},
        )
        self._event_cb = None

    def on_event(self, callback):
        """Register a callback for intermediate events."""
        self._event_cb = callback

    def run(self, goal_or_messages):
        """Send a goal string or messages array, block until result."""
        if isinstance(goal_or_messages, str):
            params = {"goal": goal_or_messages}
        else:
            params = {"messages": goal_or_messages}

        req = json.dumps({"method": "run", "params": params})
        self.proc.stdin.write(req + "\n")
        self.proc.stdin.flush()

        for line in self.proc.stdout:
            line = line.strip()
            if not line:
                continue
            event = json.loads(line)
            if self._event_cb and event["type"] not in ("result", "error"):
                self._event_cb(event)
            if event["type"] == "result":
                return event["data"]
            if event["type"] == "error":
                return event["data"]
        return None

    def close(self):
        """Terminate the subprocess."""
        if self.proc.poll() is None:
            self.proc.stdin.close()
            self.proc.wait(timeout=5)


if __name__ == "__main__":
    agent = BareAgent(provider="openai", model="gpt-4o-mini")
    print("Spawned bareagent subprocess")
    result = agent.run("What is 2 + 2?")
    if result:
        print(f"Result: {result.get('text', result)}")
    else:
        print("No result received")
    agent.close()
    print("Done")

File: contrib/ruby/bareagent.rb
# bareagent Ruby wrapper -- subprocess + JSONL over stdin/stdout.
# stdlib only: open3, json.

require "open3"
require "json"

class BareAgent
  attr_writer :on_event

  # Spawn a bareagent subprocess.
  #   BareAgent.new(provider: "openai", model: "gpt-4o-mini")
  def initialize(provider: "openai", model: nil, url: nil)
    cmd = ["npx", "bareagent", "--jsonl", "--provider", provider]
    cmd += ["--model", model] if model
    cmd += ["--url", url] if url
    @stdin, @stdout, @stderr, @wait_thr = Open3.popen3(*cmd)
    @on_event = nil
  end

  # Send a goal string or messages array, block until result.
  def run(goal_or_messages)
    params = if goal_or_messages.is_a?(String)
               { "goal" => goal_or_messages }
             else
               { "messages" => goal_or_messages }
             end

    req = JSON.generate({ "method" => "run", "params" => params })
    @stdin.puts(req)
    @stdin.flush

    @stdout.each_line do |line|
      line = line.strip
      next if line.empty?
      event = JSON.parse(line)
      if @on_event && !%w[result error].include?(event["type"])
        @on_event.call(event)
      end
      return event["data"] if %w[result error].include?(event["type"])
    end
    nil
  end

  # Terminate the subprocess.
  def close
    @stdin.close unless @stdin.closed?
    @wait_thr.join
  end
end

if __FILE__ == $PROGRAM_NAME
  agent = BareAgent.new(provider: "openai", model: "gpt-4o-mini")
  puts "Spawned bareagent subprocess"
  result = agent.run("What is 2 + 2?")
  if result
    puts "Result: #{result['text'] || result}"
  else
    puts "No result received"
  end
  agent.close
  puts "Done"
end
